[ENT-349] Display Course Info Overlay

replaceafill opened the pull request:
**Description:** This PR displays a modal dialog when the "View Course Details" link is clicked in the landing page.

**JIRA:** [ENT-349](https://openedx.atlassian.net/browse/ENT-349)

**Dependencies:** None. 

**Installation instructions:** Standard Enterprise setup

**Screenshot:**
![screenshot](https://user-images.githubusercontent.com/560781/26891889-d0839cee-4b73-11e7-805c-8e2c939346f5.png)


**Testing instructions:**

1. Add an Enterprise Customer in the Django admin. Check its ID
2. Add a course in the customer catalog. Check its ID.
3. Visit the new landing page for the enterprise customer and course by typing a URL like this:

http://localhost:8000/enterprise/7c3c286a-9834-46ae-809e-e01b114ba980/course/course-v1:edX+DemoX+Demo_Course/enroll/

where 7c3c286a-9834-46ae-809e-e01b114ba980 should be the ID of the customer and 'course-v1:edX+DemoX+Demo_Course' should be the ID of the course

4. Click the "View Course Details" in the landing page
5. Verify that a modal dialog like the screenshot above opens
6. Click the close button or outside the dialog to close it 

**Merge checklist:**

- [ ] ~Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.~
- [X] All reviewers approved
- [X] CI build is green
- [X] Version bumped
- [X] Changelog record added
- [X] Documentation updated (not only docstrings)
- [X] Commits are (reasonably) squashed
- [X] Translations are updated
- [X] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:**

1. The ENT ticket mentions several attributes to include, but I followed the image mockup as much as possible. Also the following information in the dialog has been mocked because I didn't know where to find it: 
- course level
- course length
- the what you'll learn section at the bottom

2. Also the ENT mockup has a "You Save $10.00" message in the price that I cannot generate because the original and final price attributes are just strings. So I omitted that for the moment. And for displaying the price I just got the first course mode.

**Reviewers**

- [x] @itsjeyd
- [x] @douglashall

edx-webhook commented:
Thanks for the pull request, @replaceafill! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-enterprise&number=110"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

haikuginger commented on enterprise/templates/enterprise/enterprise_course_enrollment_page.html:
```diff
@@ -88,6 +90,64 @@ <h3 class="course-confirmation-title">${ confirmation_text }</h3>
+          </div>
+          <div class="overview">
+            <h1><strong>What you'll mock</strong></h1>
+            <p>Is this what Studio calls "Course Overview"? Should this be inserted as an HTML fragment? What is the name of the attribute to look and where</p>
```
Yes, this should be an HTML fragment. I would just have a div here that can be filled into; you should be able to get this from `course_details['overview']`.

haikuginger commented on enterprise/templates/enterprise/enterprise_course_enrollment_page.html:
```diff
@@ -88,6 +90,64 @@ <h3 class="course-confirmation-title">${ confirmation_text }</h3>
+            <h1 id="modal-header-text">${ course_name }</h1>
+            <p class="short-description">${ course_short_description }</p>
+            <div class="enterprise">
+              <img src="${ enterprise_customer.branding_configuration.logo.url }" alt="${ enterprise_customer.name }" />
```
This needs to be the logo of the course organization, not the Enterprise Customer.

replaceafill commented:
@haikuginger I addressed your observations and updated the screenshot. However I'm not sure if using the branding.api to get the organization logo is the right approach. Let me know.

replaceafill commented:
@haikuginger While looking if I should pick OC-2664, I realized that both enterprise templates have the course organization logo! I'll remove the branding.api call and use the same approach.

haikuginger commented:
@replaceafill, it sounds like we just want the organization logo to be left-aligned :) So that should be simpler. Doing a more comprehensive review now.

haikuginger commented on enterprise/templates/enterprise/enterprise_course_enrollment_page.html:
```diff
@@ -88,6 +90,61 @@ <h3 class="course-confirmation-title">${ confirmation_text }</h3>
+              <li>
+                <span class="icon fa fa-clock-o" aria-hidden="true"></span>
+                <span class="title">${ duration_text}</span>
+                <span class="text">12 weeks per course</span>
```
Set this to a variable, please, and fill in your static value in the context.

haikuginger commented on enterprise/templates/enterprise/enterprise_course_enrollment_page.html:
```diff
@@ -88,6 +90,61 @@ <h3 class="course-confirmation-title">${ confirmation_text }</h3>
+              <li>
+                <span class="icon fa fa-certificate" aria-hidden="true"></span>
+                <span class="title">${ level_text}</span>
+                <span class="text">Intermediate</span>
```
So, this actually isn't a detail that's in the existing data we're using elsewhere, so I can't easily patch it in from my PR; you'll need to make a call to the Course Discovery API and use the value of the `level_type` key for the course run.

haikuginger commented on enterprise/templates/enterprise/enterprise_course_enrollment_page.html:
```diff
@@ -88,6 +90,61 @@ <h3 class="course-confirmation-title">${ confirmation_text }</h3>
+            <h1 id="modal-header-text">${ course_name }</h1>
+            <p class="short-description">${ course_short_description }</p>
+            <div class="enterprise">
+              <img src="${static.url('images/logo.png')}" alt="${ course_organization }" />
```
This still isn't correct. This is just a link to the platform logo. We need to use the logo of the _course's creating organization_.

You can fetch this information via the Organizations API with a URL like this:

```
http://localhost:8000/api/organizations/v0/organizations/edX/
```

where edX is the value of the `org` key in `course_details`.

That'll return a value like the following:

```json
{
    "id":1,
    "created":"2017-05-11T07:02:14.250525Z",
    "modified":"2017-05-11T07:02:14.308697Z",
    "name":"edX",
    "short_name":"edX",
    "description":"",
    "logo":"http://localhost:8000/media/organization_logos/Screen_Shot_2017-05-20_at_4.13.02_PM.png",
    "active":true
}
```

You can then use the details fetched using that API to select the appropriate image URL. I'd also set the alt text based on the value of the `name` key in the response.

replaceafill replied:
@haikuginger Is there any disadvantages with using the https://github.com/edx/edx-platform/blob/master/common/djangoapps/util/organizations_helpers.py module instead of the URL approach?

replaceafill commented:
@haikuginger I just addressed your observations.

itsjeyd commented:
@replaceafill I'm filling in for @haikuginger while he's away. Could you please resolve the conflicts listed above? I'll have a closer look at these changes once you've done that.

replaceafill commented:
@itsjeyd I just rebased the branch and fixed the conflicts. Let me know if you want me to squash the commits before your review.

itsjeyd commented on enterprise/static/enterprise/enterprise_course_enrollment_page.css:
```diff
@@ -208,6 +208,7 @@ body, .zebra-stripe > :nth-child(2n+1), .depth-0 {
     background: #0b5177;
 }
 
+<<<<<<< 0db5a32293f7ce4eb61133ebcc474bfe971c83c0
```
@replaceafill Leftover from conflict resolution (remove before updating sandbox).

itsjeyd commented on enterprise/static/enterprise/enterprise_course_enrollment_page.js:
```diff
@@ -0,0 +1,60 @@
+    focusableItems = container.find('*').filter(focusableElementsSelector()).filter(':visible');
+    container.attr('aria-hidden', 'true');
+    focusableItems.attr('tabindex', '-1');
+}
```
@replaceafill It might be worth DRYing out this pair of functions. Here is a rough sketch:

```js
function toggleTabIndexing(containerSelector, newState) {
    var container = $(containerSelector),
        focusableItems = container.find('*').filter(focusableElementsSelector()).filter(':visible'),
        ariaHidden = 'true',
        tabIndex = '-1';
    if (newState === 'enabled') {
        ariaHidden = 'false';
        tabIndex = '0';
    }
    container.attr('aria-hidden', ariaHidden);
    focusableItems.attr('tabindex', tabIndex);
}
function enableTabIndexingOn(containerSelector) {
    toggleTabIndexing(containerSelector, 'enabled');
}
function disableTabIndexingOn(containerSelector) {
    toggleTabIndexing(containerSelector, 'disabled');
}
```

replaceafill replied:
@itsjeyd I named it `setTabIndexing` and sent `ariaHidden` and `tabIndex` as parameters to avoid the conditional.

itsjeyd replied:
@replaceafill That's great, thanks :)

itsjeyd commented on enterprise/static/enterprise/enterprise_course_enrollment_page.js:
```diff
@@ -0,0 +1,60 @@
+    $('body').removeClass('open-modal');
+    enableTabIndexingOn('.enterprise-container');
+    $("#view-course-details-link").focus();
+}
```
@replaceafill Similar comment here: You could DRY out `showModal` and `hideModal` using the approach described above (introduce a `toggleModal` function and use value of `newState` parameter to determine behavior).

replaceafill replied:
@itsjeyd I couldn't extract this one. It would save just one line, correct?

itsjeyd replied:
@replaceafill Sorry, I should have provided more details about the approach I had in mind. Using [`toggle`](https://api.jquery.com/toggle/#toggle-display) and [`toggleClass`](https://api.jquery.com/toggleClass/), you can save more than one line:

```js
function toggleModal(modalSelector, visible) {
    var ariaHidden = 'true',
        tabIndex = '-1',
        focusSelector = '#view-course-details-link';

    if (visible) {
        ariaHidden = 'false';
        tabIndex = '0';
        focusSelector = '#modal-close-button';
    }

    $(modalSelector).toggle(visible);
    $('body').toggleClass('open-modal', visible);

    $(modalSelector).attr('aria-hidden', ariaHidden);
    $(focusSelector).focus();
    setTabIndexing('.enterprise-container', ariaHidden, tabIndex);
}
```

I would keep the conditional here (instead of introducing an `ariaHidden` parameter) to make it clear that the value of the `aria-hidden` attribute depends on the target state of the modal, and to keep control over setting that attribute inside the `toggleModal` function. When calling `toggleModal`, make sure to pass a Boolean value for `visible` (`toggleClass` requires this).

Note also that with this approach, `enableTabIndexingOn` and `disableTabIndexingOn` are no longer needed.

itsjeyd commented on enterprise/templates/enterprise/enterprise_course_enrollment_page.html:
```diff
@@ -108,6 +110,61 @@ <h3 class="course-confirmation-title">${ confirmation_text }</h3>
+                <span class="icon fa fa-tag" aria-hidden="true"></span>
+                <span class="title">${ price_text}</span>
+                <span class="text">
+                  % if course_modes:
```
@replaceafill The code below accesses `course_modes[0]`, `course_modes[0]['final_price']`, and `course_modes[0]['original_price']` several times, so it would make sense to introduce variables that reference these values. (See http://docs.makotemplates.org/en/latest/syntax.html#python-blocks for an example.)

replaceafill replied:
@itsjeyd I did the same but using django templates.

itsjeyd commented on enterprise/views.py:
```diff
@@ -477,6 +482,10 @@ class CourseEnrollmentView(View):
         'continue_link_text': _('Continue'),
+        'level_text': _('Level'),
+        'effort_text': _('Effort'),
+        'effort': _('{hours} hours per week, per course'),
```
@replaceafill IIUC, `effort` is used to generate `course_effort` below. You could rename it to `course_effort_text` to be more precise, and to stick with the `*_text` naming scheme that other context variables are using.

itsjeyd commented on tests/test_views.py:
```diff
@@ -971,8 +973,12 @@ def _login(self):
-    def test_get_course_enrollment_page(
+    @mock.patch('enterprise.views.organizations_helpers')
+    @mock.patch('enterprise.views.CourseCatalogApiClient')
+    def test_get_course_specific_consent(
```
@replaceafill I might be missing something, but why did you rename this test and the one below?

replaceafill replied:
@itsjeyd Unintended. This is me not being good at rebasing with git (or git at all...). Thanks for catching it.

replaceafill commented:
@itsjeyd I addressed most of your observations and left a couple of comments. Please take another look.

itsjeyd commented on enterprise/templates/enterprise/enterprise_course_enrollment_page.html:
```diff
@@ -45,15 +50,15 @@ <h3 class="course-confirmation-title">{{ confirmation_text }}</h3>
             <div class="radio">
               {% if course_modes|length > 1 %}
-                <input type="radio" name="data_sharing_consent" id="radio${{ forloop.counter0 }}" {% if forloop.first %}checked="checked"{% endif %} />
+                <input type="radio" name="data_sharing_consent" id="radio{{ forloop.counter0 }}" {% if forloop.first %}checked="checked"{% endif %} />
```
@replaceafill Did you check whether these IDs are referenced somewhere else? Just asking because if they are, removing `$` might break existing functionality.

replaceafill replied:
@itsjeyd Yes. I checked the CSS file just in case. I think that sign is just a left over from the mako to django template conversion.

itsjeyd commented on tests/test_views.py:
```diff
@@ -1281,3 +1316,193 @@ def test_get_course_specific_consent_for_enrolled_user(
+            social_auth_object_mock,   # pylint: disable=unused-argument
+            get_ec_for_request_mock,   # pylint: disable=unused-argument
+            configuration_helpers_mock,
+            render_to_response_mock,    # pylint: disable=unused-argument
```
@replaceafill Instead of adding `# pylint: disable=unused-argument` to all of these, you can add `_unused` as a suffix to the parameter names, see the definition of [`dummy-variables-rgx`](https://github.com/edx/edx-enterprise/blob/master/pylintrc#L155) for this repo.

itsjeyd commented on tests/test_views.py:
```diff
@@ -1281,3 +1316,193 @@ def test_get_course_specific_consent_for_enrolled_user(
+            'close_modal_button_text': 'Close',
+        }
+        for key, value in expected_context.items():
+            assert response.context[key] == value  # pylint: disable=no-member
```
@replaceafill Thanks for adding these tests!

There's a lot of duplication between them. Aside from the method name, the only lines that differ are:

```python
# From test_get_course_enrollment_page_with_no_course_effort:
        dummy_demo_course_details_data['effort'] = None
# From test_get_course_enrollment_page_with_unparseable_course_effort:
        dummy_demo_course_details_data['effort'] = 'Cannot convert to integer'
```

Could you DRY them out using `ddt`?

Also, if you could add some inline comments about what's happening (as well as a bit of vertical whitespace to group logical units), that would make the code easier to digest.

douglashall commented:
@replaceafill Please rebase now that https://github.com/edx/edx-enterprise/pull/109 has merged.

replaceafill commented:
@douglashall @itsjeyd I just rebased this on master (ce7b16e).

itsjeyd commented on enterprise/templates/enterprise/enterprise_course_enrollment_page.html:
```diff
@@ -79,4 +84,62 @@ <h3 class="course-confirmation-title">{{ confirmation_text }}</h3>
+          </ol>
+        </div>
+        <div class="overview">
+          {{ course_overview|safe }}
```
@replaceafill The course overview looks good, but it would be great if we could tweak the styling for section headers ("About This Course", "Prerequisites", etc. for the demo course) to match the mock-ups even more closely. According to the mockup they should be bold and using a font that's slightly larger than the text below them.

replaceafill commented:
@itsjeyd I just addressed the TAB issue and adjusted the styles of the headers in the overview section. I also handled missing organizations and course runs better.

itsjeyd commented on enterprise/views.py:
```diff
@@ -590,9 +591,16 @@ def get_enterprise_course_enrollment_page(self, request, enterprise_customer, co
 
+        try:
+            organization = organizations_helpers.get_organization(course_details['org'])
+            organization_logo = organization['logo'].url
```
@replaceafill There's one more edge case here that we should probably handle: To test this change I deleted the logo from the organization I had created in the Django admin (by checking "clear" next to the current logo and clicking "Save"). When I tried to access the landing page after that this line errored out with:

> `ValueError`: The 'logo' attribute has no file associated with it.

itsjeyd commented on tests/test_views.py:
```diff
@@ -1772,3 +1772,111 @@ def test_get_course_enrollment_page_with_unparseable_course_effort(
+
+        # Compare expected context with response result
+        for key, value in expected_context.items():
+            assert response.context[key] == value  # pylint: disable=no-member
```
@replaceafill There's quite a bit of duplication between these tests (and in this file in general). Just wanted to point this out -- since we need to move forward with this, I'm fine with deferring further work on refactoring the tests to another PR/ticket.

itsjeyd commented on tests/test_views.py:
```diff
@@ -1880,3 +1880,111 @@ def test_get_course_enrollment_page_with_no_organization(
+
+        # Compare expected context with response result
+        for key, value in expected_context.items():
+            assert response.context[key] == value  # pylint: disable=no-member
```
@replaceafill This test is pretty much an exact copy of `test_get_course_enrollment_page_with_no_organization` -- please use `ddt` to reduce duplication.

brittneyexline commented on enterprise/views.py:
```diff
@@ -597,6 +626,16 @@ def get_enterprise_course_enrollment_page(self, request, enterprise_customer, co
+            'organization_logo': organization_logo,
+            'organization_name': organization_name,
+            'duration': duration,
+            'course_level_type': course_run.get('level_type', ''),
```
I believe there is a level type at the course level - this seems to be the only field we're pulling out of the course run right now so it would be nice to avoid that extra call.

replaceafill replied:
@brittneyexline I couldn't find a relevant key in the course details retrieved with the enterprise.lms_api.CourseApiClient (or the serializer in the lms.djangoapps.course_api package).

brittneyexline commented on enterprise/templates/enterprise/enterprise_course_enrollment_page.html:
```diff
@@ -79,4 +84,62 @@ <h3 class="course-confirmation-title">{{ confirmation_text }}</h3>
+                      <br />
+                      <span class="discount">{{ course_mode.final_price }}</span>
+                    {% else %}
+                      {{ course_mode.original_price }}
```
@chris-mike do you have any input on what we should do if there are no course modes for the course? This affects the price we display. There are also similar questions below around what to do if the other information around level and effort are missing in the metadata.

chris-mike replied:
Just so I'm clear, when you say "course modes" you're asking: what happens to price if there is no price? These 4 items should be available for most courses. In the case they are not, we could go with Marketing defaults. I'll reach out to them to see if there's a different approach for enterprise (particularly as it relates to pricing). Example, it could be "contact your administrator" or something along those lines.

chris-mike commented:
@replaceafill @itsjeyd @brittneyexline 
In cases where there is no data, the line should not be displayed at all.  This is what Marketing does with price.

To get this over the line, we could insert a "---", but the line disappearing would be preferable.

grantgoodmanedX commented:
@chris-mike I agree that my first choice would be to not display the line at all, but if we must, use "---" or something similar to indicate that there is no data.

replaceafill commented:
Thanks @chris-mike @grantgoodmanedX The metadata rows are not displayed if there's no data now:

![empty](https://user-images.githubusercontent.com/560781/26949035-a4f56002-4c55-11e7-8321-8ac83ec4fd7f.png)

douglashall commented:
@replaceafill Can you squash this in preparation for merge?

replaceafill commented:
@douglashall Squashed.

itsjeyd commented:
@replaceafill Just a quick note on squashing: Next time you do that, can you please post a comment that includes the hash of the commit that was at the tip of your branch *before* squashing? That is, instead of just saying

> Squashed.

... say something like

> Squashed from <hash>.

GitHub will automatically create a working link to the commit identified by `<hash>`. That makes it possible (or at least easier) for your reviewers to find individual commits (that they haven't reviewed) later on.

I'm going to go ahead and update your comment with the hash of (what I think is) the most recent commit now.

itsjeyd commented on tests/test_views.py:
```diff
@@ -974,6 +979,56 @@ def _login(self):
+            'level_type': 'Type 1',
+        }
+
+    def _setup_organizations_client(self, client_mock, data=DUMMY_PLACEHOLDER, side_effect=None):
```
@replaceafill There's only one instance where client code is passing custom values for `data` and `side_effect` to this method: https://github.com/replaceafill/edx-enterprise/blob/d7ef90e2816ec74d997f2dad09e39d62aef8ead6/tests/test_views.py#L1806. I think you should move any custom logic there, and only call this method from tests that need an identical setup for the organizations client. That will allow you to get rid of the `data` and `side_effect` parameters here:

```python
    def _setup_organizations_client(self, client_mock):
        """
        Sets up the Organizations API client
        """
        logo_mock = mock.MagicMock()
        logo_mock.url = 'logo.png'
        data = {
            'logo': logo_mock,
            'name': 'Organization',
        }
        client_mock.get_organization.return_value = data
```

Below:

```python
    import inspect

    # ...
    @ddt.data(None, ValueError)
    def test_get_course_enrollment_page_organization_errors(
            self,
            organizations_data,
            # ...
            organizations_helpers_mock,
            # ...
    ):
        # ...

        # Set up organizations API client
        if inspect.isclass(organizations_data) and issubclass(organizations_data, Exception):
            organizations_helpers_mock.get_organization.side_effect = organizations_data
        else:
            organizations_helpers_mock.get_organization.return_value = organizations_data
```
```

itsjeyd commented on tests/test_views.py:
```diff
@@ -974,6 +979,56 @@ def _login(self):
+        else:
+            client_mock.get_organization.side_effect = side_effect
+
+    def _setup_course_api_client(self, client_mock, data=DUMMY_PLACEHOLDER, side_effect=None):
```
@replaceafill Similar comment here: There's 13 calls to this method but only a few of them pass custom values for `data` and/or `side_effect`. You can remove these calls and specify custom behavior inline; that will let you get rid of the `data` and `side_effect` parameters here.

itsjeyd commented on tests/test_views.py:
```diff
@@ -974,6 +979,56 @@ def _login(self):
+        else:
+            client.get_course_details.side_effect = HttpClientError
+
+    def _setup_enrollment_client(self, client_mock, data=DUMMY_PLACEHOLDER, side_effect=None):
```
@replaceafill Similar comment here: There's 10 calls to this method but only a couple pass custom values for `data` and/or `side_effect`. You can remove these calls and specify custom behavior inline; that will let you get rid of the `data` and `side_effect` parameters here.

Ultimately this will also let you get rid of `DUMMY_PLACEHOLDER` altogether.

douglashall commented on enterprise/views.py:
```diff
@@ -572,6 +583,24 @@ def get_enterprise_course_enrollment_page(self, request, enterprise_customer, co
         if course_details['start']:
             course_start_date = parse(course_details['start']).strftime('%B %d, %Y')
 
+        duration = _('12 weeks per course')
```
Let's remove duration for now. I will create a followup story for implementing it dynamically.

douglashall commented on enterprise/views.py:
```diff
@@ -505,6 +511,11 @@ class CourseEnrollmentView(View):
+        'level_text': _('Level'),
+        'effort_text': _('Effort'),
+        'effort_hours_text': _('{hours} hours per week, per course'),
+        'duration_text': _('Average Length'),
```
You can remove this as well.

itsjeyd commented:
Thanks for the review, @douglashall! Do we need another (explicit) 👍 before we can merge this? CC @brittneyexline

itsjeyd commented:
@haikuginger Dismissing your (now outdated) review as discussed internally.

douglashall commented:
Awesome! Let's bump the package version, squash, and merge!

replaceafill commented:
@douglashall @itsjeyd I went through the merge checklist and rebased/squashed this from 490cedf

itsjeyd commented:
Thanks @replaceafill! The additional changes that you made look good.

@brittneyexline Do you want to give this a quick once-over as well?

itsjeyd commented:
Thanks @brittneyexline! Merging now.
